netket/machine/jax.py

```diff
@@ -347,6 +347,165 @@ def JaxRbm(hilbert, alpha, dtype=complex):
     )
 
 
+def MPSPeriodic(
+    hilbert, bond_dim, diag=False, symperiod=-1, dtype=complex, outdtype=complex
+):
+    r"""
+    Constructs a periodic Matrix Product State (MPS) for a quantum state of discrete
+    degrees of freedom, wrapped as Jax machine.  The MPS is defined as
+
+    .. math:: \Psi(s_1,\dots s_N) = \Tr(A[s_1]\dots A[s_N]),
+
+    for arbitrary local quantum numbers :math:`s_i`, where :math:`A[s_1]` is a matrix
+    of dimension (bdim,bdim), depending on the value of the local quantum number :math:`s_i`.
+
+        Args:
+            hilbert: Hilbert space on which the state is defined. Should be a
+                subclass of `netket.hilbert.Hilbert`.
+            bond_dim (int): Virtual dimension of the MPS tensors.
+            diag (bool): Whether or not to use diagonal matrices in the MPS tensors.
+                default=False
+            symperiod (int): Periodicity in the chain of MPS tensors. For positive
+                values, the chain of MPS tensors is constructed as a sequence of
+                identical unit cells consisting of symperiod tensors. For negative
+                values, the period is equal to the number of local physical variables.
+                default=-1
+            dtype: complex or float, whether the variational parameters of the MPS
+                are real or complex. default=complex
+            outdtype: complex or float, whether the scalar output of the MPS is
+                real or complec. default=complex
+
+        returns:
+            Jax machine of the Matrix Product state.
+    """
+    return Jax(
+        hilbert,
+        stax.serial(MpsPeriodicLayer(hilbert, bond_dim, diag, symperiod, dtype)),
+        dtype=dtype,
+        outdtype=outdtype,
```

Comment:
basically the solution would be to remove outdtype here and pass only dtype to ```machine.Jax```

Reply:
Thanks, did that in the last commit.  I noticed that putting `dtype=float` yields the following error: `TypeError: Type of cotangent input to vjp pullback function (float64) does not match type of corresponding primal output (complex128)`.  I think this is due to the dtype of `params` changing after the first iteration.  The gradients used to update params have `complex128` dtype, causing the `params` to be cast to `complex128` on the first parameter update.  This leads to inconsistencies on the second iteration as the output and parameters of the network are now `complex128`.  This seems to be a broader issue because it also occurs using the `JaxRbm` machine.  I think forcing the dtype to be `float64` in the definition of `_vjp_fun` for the R->R case should resolve this issue.

Reply:
I think @gcarleo fixed this issue in one of the last PRs on Saturday.
Is this still causing issues?

Reply:
yes, this happened with the PR of Saturday.

Reply:
I think the error arises because your machine always outputs a complex number, even if `outdtype=float`.

Reply:
Here is the evil guy.
Removing this line will solve this problem
https://github.com/netket/netket/blob/35536bccd4cbfb1c86eaffd2154d6ebe218b1ad9/netket/_vmc.py#L165

Reply:
If you rebase on top of #430 this should be fixed.